source/mir/math/stat.d

```diff
@@ -25,25 +25,53 @@ import mir.primitives;
 import std.range.primitives: isInputRange;
 import std.traits: isArray, isFloatingPoint;
 
+private U prod(T : U[N], U, size_t N)(T input)
+    if (N > 0)
+{
+    static if (N == 1) {
+        return input[0];
+    } else {
+        U output = input[0];
+        for (size_t i = 1; i < N; i++)
+            output *= input[i];
+        return output;
+    }
+}
+
+version(mir_test) @safe pure nothrow @nogc unittest
+{
+    size_t[1] x = [3];
+    size_t[2] y = [4, 2];
+    
+    assert(x.prod == 3);
+    assert(y.prod == 8);
+}
+
 /++
-Computes the average of `r`, which must be a finite iterable.
+Computes the arithmetic mean of `r`, which must be a finite iterable.
 
 Returns:
-    The average of all the elements in the range r.
+    The arithmetic mean of all the elements in the range r.
 +/
 template mean(Summation summation = Summation.appropriate)
 {
     ///
     @safe @fmamath sumType!Range
     mean(Range)(Range r)
-        if (hasLength!Range
+        if (hasShape!Range
+         || hasLength!Range
          || summation == Summation.appropriate
          || summation == Summation.fast
          || summation == Summation.naive)
     {
-        static if (hasLength!Range)
+        static if (hasShape!Range || hasLength!Range)
         {
-            auto n = r.length;
+            static if (hasShape!Range)
+            {
+                auto n = r.shape.prod;
+            } else {
+                auto n = r.length;
+            }
             return sum!summation(r.move) / cast(sumType!Range) n;
         }
         else
```

Comment:
this branch seems never used.

Reply:
That branch was in there previously. In the original code, the template constraints were hasLength or several summation checks. When hasLength == false, then it used this as a fallback. I don't know if it pulls its weight, but if it is removed, then the template constraints should be simplified.

Reply:
Then the `Summator` approach is preferred instead of seed in the function API.
see also https://github.com/libmir/mir-algorithm/pull/238#issuecomment-618158406